toc.md

```diff
@@ -650,42 +818,32 @@ module Workerholic
 
-#### Historical Statistics
-Live data, check! Next, we want to display historical, and first we had to think about what type of data to store?
+Initially, we decided not to store live data and to just poll new data every 10 seconds. Once we introduced graphs into the web UI just polling for new data and throwing away stale data was no longer an option.
 
+At first, we decided to store data on the front-end for a certain number of data points so we could generate graphs. This worked, but only if the user stayed on the page. If the user navigated away from the page, the data be lost. Instead, we looked at Redis to store this data, up to 1,000 seconds, so we could have a total of 100 data points.
+
+Note that we also do some cleanup and remove the jobs that exceed this 1,000 seconds time range from our `workerholic:stats` sorted sets.
+
+After solving the problem for live data we decided to introduce historical data so that uses could have a broader view of their completed/failed jobs.
+
+#### Historical Statistics
 ##### What data?
-![reporting_historical_charts](/images/reporting_historical_charts.png)
 
-Since historical data is looking into the past, for now we decided that we'll just store aggregated data of completed and failed jobs, as well as the breakdown for each class, up to 365 days.
+We decided to display data for completed and failed jobs, as well as the breakdown for each class, up to 365 days. This way users would have a broad overview of their background jobs state over time.
+
+![reporting_historical_charts](/images/reporting_historical_charts.png)
 
 ##### How to store the data?
 ###### First Iteration
-```ruby
-module Workerholic
-  class Storage
-    # ...
-
-    def self.save_job(category, job)
-      job_hash = job.to_hash
-      serialized_job_stats = JobSerializer.serialize(job_hash)
 
-      namespace = "workerholic:stats:#{category}:#{job.klass}"
-      storage.add_to_set(namespace, job.statistics.completed_at, serialized_job_stats)
-    end
-
-    # ...
-  end
-end
-```
+First of all, we decided to do store some aggregated data instead of serialized jobs, this way it would take much less space in Redis.
 
-In our first iteration, we used a sorted set using the beginning of the day as a timestamp to use as scores, which would be very easy to retrieve from Redis - using a range of scores to display data for 7 or 30 days for example. However, we ran into concurrency issues because we have three ways to update aggregated data: getting the count, removing the count, and incrementing the count.
+In our first iteration, we used a sorted set using the first timestamp of the day as a score. This approach would make it easy to retrieve a specific data range from Redis.
```

Comment:
In our first iteration, we used a sorted set with the first timestamp of the day as a score. This approach would make it easy to retrieve specific data range from Redis.